modules/sr/robot3/arduino_devices.py

```diff
@@ -0,0 +1,234 @@
+class PinDevice(abc.ABC):
+    """
+    A pin on the Arduino.
+    """
+    __slots__ = ('_supports_analogue', '_disabled', '_mode')
+
+    _ANALOGUE_RANGE = (0., 5.)  # Volts
+
+    def __init__(
+        self,
+        supports_analogue: bool,
+        disabled: bool = False,
+    ) -> None:
+        """
+        :param supports_analogue: Whether the pin supports analogue reads.
+        :param disabled: Whether the pin can be controlled.
+        """
+        self._supports_analogue = supports_analogue
+        self._disabled = disabled
+        self._mode = GPIOPinMode.INPUT
+
+    def _check_if_disabled(self) -> None:
+        if self._disabled:
+            raise IOError('This pin cannot be controlled.')
+
+    @property
+    def mode(self) -> GPIOPinMode:
+        """
+        Get the mode of the pin.
+
+        This returns the cached value since the board does not report this.
+
+        :raises IOError: If this pin cannot be controlled.
+        :return: The mode of the pin.
+        """
+        self._check_if_disabled()
+        return self._mode
+
+    @mode.setter
+    def mode(self, value: GPIOPinMode) -> None:
+        """
+        Set the mode of the pin.
+
+        To do analogue or digital reads set the mode to INPUT or INPUT_PULLUP.
+        To do digital writes set the mode to OUTPUT.
+
+        :param value: The mode to set the pin to.
+        :raises IOError: If the pin mode is not a GPIOPinMode.
+        :raises IOError: If this pin cannot be controlled.
+        """
+        self._check_if_disabled()
+        if not isinstance(value, GPIOPinMode):
+            raise IOError('Pin mode only supports being set to a GPIOPinMode')
+
+        self._mode = value
+
+    def _digital_read(self) -> bool:
+        return self._analogue_read() > 1
+
+    def digital_read(self) -> bool:
+        """
+        Perform a digital read on the pin.
+
+        :raises IOError: If the pin's current mode does not support digital read
+        :raises IOError: If this pin cannot be controlled.
+        :return: The digital value of the pin.
+        """
+        self._check_if_disabled()
+        if self.mode not in DIGITAL_READ_MODES:
+            raise IOError(f'Digital read is not supported in {self.mode}')
+        return self._digital_read()
+
+    def _digital_write(self, value: bool) -> None:
+        return
+
+    def digital_write(self, value: bool) -> None:
+        """
+        Write a digital value to the pin.
+
+        :param value: The value to write to the pin.
+        :raises IOError: If the pin's current mode does not support digital write.
+        :raises IOError: If this pin cannot be controlled.
+        """
+        self._check_if_disabled()
+        if self.mode not in DIGITAL_WRITE_MODES:
+            raise IOError(f'Digital write is not supported in {self.mode}')
+
+        self._digital_write(value)
+
+    def _analogue_read(self) -> float:
+        return map_to_range(0, 1, *self._ANALOGUE_RANGE, random.random())
```

Comment:
Why is this returning a random number? Also the splat operation makes it confusing to line up the arguments to the function.

Reply:
This is sort-of attempting to emulate a floating pin.

Reply:
This isn't really necessary and is going to cause confusion, just setting it to 0 is adequate.

Reply:
Perhaps we should return `nan` here or something? What would a real arduino do if we tried to read a value from a floating pin?

Reply:
> Also the splat operation makes it confusing to line up the arguments to the function.

https://github.com/srobo/competition-simulator/pull/423 should help with that (it's always felt odd to me that the values passed to that util weren't tuples given the nature of the couplings). I'd had in mind to do this while working on this; the splat was a temporary thing to allow that change to be a separate PR.